### PR TITLE
[SPARK-48380][CORE][WIP]: SerDeUtil.javaToPython to support rows up to 2GB

### PR DESCRIPTION
**Pull request by zshao9**
### What changes were proposed in this pull request?
This simple PR changes the behavior of AutoBatchPickler to be able to catch situations where a single row is smaller than 2GB, but a batch of them exceeds 2GB and cause issues in underlying buffer management.

### Why are the changes needed?
With this, pyspark program can decide to use a different batch size when the default AutoBatchPickler ends up with 2GB exceeded error as in the JIRA ticket: https://issues.apache.org/jira/browse/SPARK-48380

### Does this PR introduce _any_ user-facing change?
No.  This code is entirely transparent to users.

```

### How was this patch tested?
```
> build/sbt -mem 32000
sbt:spark-parent> project core
sbt:spark-core> testOnly org.apache.spark.api.python.SerDeUtilSuite
```

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/api/python/SerDeUtil.scala [hunk 1]
```diff
@@ -104,12 +104,37 @@ private[spark] object SerDeUtil extends Logging {
     }
   }
 
+  /**
+   * Use a fixed batch size
+   */
+  private[spark] class FixedBatchedPickler(iter: Iterator[Any], batch: Integer) extends Iterator[Array[Byte]] {
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/api/python/SerDeUtil.scala:110]:
```suggestion
  private[spark] class FixedBatchedPickler(iter: Iterator[Any], batch: int) extends Iterator[Array[Byte]] {
```

File: core/src/main/scala/org/apache/spark/api/python/SerDeUtil.scala [hunk 2]
```diff
@@ -104,12 +104,37 @@ private[spark] object SerDeUtil extends Logging {
...
+
   /**
    * Convert an RDD of Java objects to an RDD of serialized Python objects, that is usable by
    * PySpark.
+   * @param batchSize if non-zero, the fixed batch size to use when pickling objects
    */
-  def javaToPython(jRDD: JavaRDD[_]): JavaRDD[Array[Byte]] = {
-    jRDD.rdd.mapPartitions { iter => new AutoBatchedPickler(iter) }
+  def javaToPython(jRDD: JavaRDD[_], batchSize: Integer = 0): JavaRDD[Array[Byte]] = {
+    if (batchSize == 0) {
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/api/python/SerDeUtil.scala:133]: shall we fail if `batchSize < 0`?

File: core/src/main/scala/org/apache/spark/api/python/SerDeUtil.scala [hunk 3]
```diff
@@ -104,12 +104,37 @@ private[spark] object SerDeUtil extends Logging {
...
+  }
+
   /**
    * Convert an RDD of Java objects to an RDD of serialized Python objects, that is usable by
    * PySpark.
+   * @param batchSize if non-zero, the fixed batch size to use when pickling objects
    */
-  def javaToPython(jRDD: JavaRDD[_]): JavaRDD[Array[Byte]] = {
-    jRDD.rdd.mapPartitions { iter => new AutoBatchedPickler(iter) }
+  def javaToPython(jRDD: JavaRDD[_], batchSize: Integer = 0): JavaRDD[Array[Byte]] = {
```
[the remark this answers is not in the document]
Comment (zshao9) [reply core/src/main/scala/org/apache/spark/api/python/SerDeUtil.scala]: Will do

File: core/src/main/scala/org/apache/spark/api/python/SerDeUtil.scala (hunk 1 above)
Comment (zshao9) [reply core/src/main/scala/org/apache/spark/api/python/SerDeUtil.scala]: Will do!

File: core/src/test/scala/org/apache/spark/api/python/SerDeUtilSuite.scala [hunk 4]
```diff
@@ -32,5 +32,22 @@ class SerDeUtilSuite extends SparkFunSuite with SharedSparkContext {
...
+    val collected = pythonRdd.collect()
+    assert(collected.size() === 3)
+  }
+
+  test("Converting JavaRDD to PythonRDD with an auto batch size (SPARK-48380)") {
+    val rddWith3Elements = sc.makeRDD(Seq[(Any, Any)]((1, 2), (3, 4), (5, 6)))
+    val javaRdd = rddWith3Elements.toJavaRDD()
+    val pythonRdd = SerDeUtil.javaToPython(javaRdd)
+    val collected = pythonRdd.collect()
+    assert(collected.size() === 2)
```
Comment (zshao9) [inline core/src/test/scala/org/apache/spark/api/python/SerDeUtilSuite.scala:49]: This is actually failing.  Need to find a way to verify that AutoBatchedPickler did batch rows together.

File: core/src/main/scala/org/apache/spark/api/python/SerDeUtil.scala [hunk 5]
```diff
@@ -104,12 +104,40 @@ private[spark] object SerDeUtil extends Logging {
     }
   }
 
+  /**
+   * Use a fixed batch size
+   */
+  private[spark] class FixedBatchedPickler(iter: Iterator[Any], batch: Int) extends Iterator[Array[Byte]] {
```
[the remark this answers is not in the document]
Comment (zshao9) [reply core/src/main/scala/org/apache/spark/api/python/SerDeUtil.scala]: Fixed

File: core/src/test/scala/org/apache/spark/api/python/SerDeUtilSuite.scala (hunk 4 above)
Comment (zshao9) [reply core/src/test/scala/org/apache/spark/api/python/SerDeUtilSuite.scala]: Fixed by using a different test

**Pull request closed without merging**